[BUG] Cython master crashes analysing division of int by float

### Describe the bug

I have bisected this problem to 0adbcde2377c2712b833f697a5ee5c64490fdeee

I just tried to build python-flint using Cython master but it crashes with
```console
$ meson compile -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /home/oscar/.pyenv/shims/ninja -C /home/oscar/current/active/python-flint/build
ninja: Entering directory `/home/oscar/current/active/python-flint/build'
[1/99] Compiling Cython source /home/oscar/current/active/python-flint/src/flint/types/fmpz.pyx
FAILED: src/flint/types/fmpz.cpython-311-x86_64-linux-gnu.so.p/src/flint/types/fmpz.pyx.c 
cython -M --fast-fail -3 -X embedsignature=True -X emit_code_comments=True -X linetrace=True /home/oscar/current/active/python-flint/src/flint/types/fmpz.pyx -o src/flint/types/fmpz.cpython-311-x86_64-linux-gnu.so.p/src/flint/types/fmpz.pyx.c

Error compiling Cython file:
------------------------------------------------------------
...
cdef inline long prec_to_dps(n):
    return max(1, int(round(int(n)/3.3219280948873626)-1))
                                  ^
------------------------------------------------------------

/home/oscar/current/active/python-flint/src/flint/utils/conversion.pxd:2:34: Compiler crash in AnalyseExpressionsTransform

ModuleNode.body = StatListNode(conversion.pxd:1:0)
StatListNode.stats[0] = CFuncDefNode(conversion.pxd:1:5,
    args = [...]/1,
    inline_in_pxd = True,
    modifiers = [...]/1,
    outer_attrs = [...]/2,
    visibility = 'private')
File 'Nodes.py', line 390, in analyse_expressions: StatListNode(conversion.pxd:2:4,
    is_terminator = True)
File 'Nodes.py', line 6807, in analyse_expressions: ReturnStatNode(conversion.pxd:2:4,
    is_terminator = True)
File 'UtilNodes.py', line 282, in analyse_types: EvalWithTempExprNode(conversion.pxd:2:21,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 6080, in analyse_types: SimpleCallNode(conversion.pxd:2:21,
    analysed = True,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 8447, in analyse_types: TupleNode(conversion.pxd:2:21,
    is_sequence_constructor = 1,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 11594, in analyse_types: SubNode(conversion.pxd:2:54,
    infix = True,
    operator = '-',
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 6080, in analyse_types: SimpleCallNode(conversion.pxd:2:27,
    analysed = True,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 8447, in analyse_types: TupleNode(conversion.pxd:2:27,
    is_sequence_constructor = 1,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 11596, in analyse_types: DivNode(conversion.pxd:2:34,
    ctruedivision = True,
    infix = True,
    operator = '/',
    result_is_used = True,
    truedivision = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 12180, in analyse_operation: DivNode(conversion.pxd:2:34,
    ctruedivision = True,
    infix = True,
    operator = '/',
    result_is_used = True,
    truedivision = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 11609, in analyse_operation: DivNode(conversion.pxd:2:34,
    ctruedivision = True,
    infix = True,
    operator = '/',
    result_is_used = True,
    truedivision = True,
    use_managed_ref = True)

Compiler crash traceback from this point on:
  File "/home/oscar/.pyenv/versions/3.11.3/envs/python-flint-3.11/lib/python3.11/site-packages/Cython/Compiler/ExprNodes.py", line 11609, in analyse_operation
    assert self.type.is_pyobject
AssertionError: 
[2/99] Compiling Cython source /home/oscar/current/active/python-flint/src/flint/pyflint.pyx
FAILED: src/flint/pyflint.cpython-311-x86_64-linux-gnu.so.p/src/flint/pyflint.pyx.c 
cython -M --fast-fail -3 -X embedsignature=True -X emit_code_comments=True -X linetrace=True /home/oscar/current/active/python-flint/src/flint/pyflint.pyx -o src/flint/pyflint.cpython-311-x86_64-linux-gnu.so.p/src/flint/pyflint.pyx.c

Error compiling Cython file:
------------------------------------------------------------
...
cdef inline long prec_to_dps(n):
    return max(1, int(round(int(n)/3.3219280948873626)-1))
                                  ^
------------------------------------------------------------

/home/oscar/current/active/python-flint/src/flint/utils/conversion.pxd:2:34: Compiler crash in AnalyseExpressionsTransform
```

### Code to reproduce the behaviour:

I haven't tried distilling this down to a standalone reproducer but the code being complained about is:
```cython
cdef inline long prec_to_dps(n):
    return max(1, int(round(int(n)/3.3219280948873626)-1))
```


### Expected behaviour

This all builds fine with Cython 3.0.10 so I expected it to build fine.

### OS

Linux

### Python version

3.12

### Cython version

master

### Additional context

_No response_

Here is a simple reproducer:
```console
$ cat tmp.pyx 
cdef inline long prec_to_dps(n):
    return max(1, int(round(int(n)/3.3219280948873626)-1))
$ cython tmp.pyx 
/home/oscar/.pyenv/versions/3.12.0/envs/sympy-3.12.git/lib/python3.12/site-packages/Cython/Compiler/Main.py:373: FutureWarning: Cython directive 'language_level' not set, using '3' (Py3). This has changed from earlier releases! File: /home/oscar/current/active/sympy/tmp.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)

Error compiling Cython file:
------------------------------------------------------------
...
cdef inline long prec_to_dps(n):
    return max(1, int(round(int(n)/3.3219280948873626)-1))
                                  ^
------------------------------------------------------------

tmp.pyx:2:34: Compiler crash in AnalyseExpressionsTransform

ModuleNode.body = StatListNode(tmp.pyx:1:5)
StatListNode.stats[0] = CFuncDefNode(tmp.pyx:1:5,
    args = [...]/1,
    modifiers = [...]/1,
    outer_attrs = [...]/2,
    visibility = 'private')
File 'Nodes.py', line 390, in analyse_expressions: StatListNode(tmp.pyx:2:4,
    is_terminator = True)
File 'Nodes.py', line 6858, in analyse_expressions: ReturnStatNode(tmp.pyx:2:4,
    is_terminator = True)
File 'UtilNodes.py', line 282, in analyse_types: EvalWithTempExprNode(tmp.pyx:2:21,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 6155, in analyse_types: SimpleCallNode(tmp.pyx:2:21,
    analysed = True,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 8661, in analyse_types: TupleNode(tmp.pyx:2:21,
    is_sequence_constructor = 1,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 11831, in analyse_types: SubNode(tmp.pyx:2:54,
    infix = True,
    operator = '-',
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 6155, in analyse_types: SimpleCallNode(tmp.pyx:2:27,
    analysed = True,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 8661, in analyse_types: TupleNode(tmp.pyx:2:27,
    is_sequence_constructor = 1,
    result_is_used = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 11833, in analyse_types: DivNode(tmp.pyx:2:34,
    ctruedivision = True,
    infix = True,
    operator = '/',
    result_is_used = True,
    truedivision = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 12417, in analyse_operation: DivNode(tmp.pyx:2:34,
    ctruedivision = True,
    infix = True,
    operator = '/',
    result_is_used = True,
    truedivision = True,
    use_managed_ref = True)
File 'ExprNodes.py', line 11846, in analyse_operation: DivNode(tmp.pyx:2:34,
    ctruedivision = True,
    infix = True,
    operator = '/',
    result_is_used = True,
    truedivision = True,
    use_managed_ref = True)

Compiler crash traceback from this point on:
  File "/home/oscar/.pyenv/versions/3.12.0/envs/sympy-3.12.git/lib/python3.12/site-packages/Cython/Compiler/ExprNodes.py", line 11846, in analyse_operation
    assert self.type.is_pyobject
AssertionError: 
```

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -12420,6 +12420,13 @@ class DivNode(NumBinopNode):
     def analyse_operation(self, env):
         self._check_truedivision(env)
         result = NumBinopNode.analyse_operation(self, env)
+
+        # The assumption here is that result is either 'self' or a coercion
+        # node containing 'self'. Thus it is reasonable to keep manipulating
+        # 'self' even if it's been replaced as the eventual result.
+        from . import Visitor
+        assert Visitor.tree_contains(result, self)
+
         if self.is_cpp_operation():
             self.cdivision = True
         if not self.type.is_pyobject:

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -11830,8 +11830,7 @@ class BinopNode(ExprNode):
     def analyse_types(self, env):
         self.operand1 = self.operand1.analyse_types(env)
         self.operand2 = self.operand2.analyse_types(env)
-        self.analyse_operation(env)
-        return self
+        return self.analyse_operation(env)
 
     def analyse_operation(self, env):
         if self.is_pythran_operation(env):
@@ -11843,12 +11842,17 @@ class BinopNode(ExprNode):
             self.coerce_operands_to_pyobjects(env)
             self.type = self.result_type(self.operand1.type,
                                          self.operand2.type, env)
-            assert self.type.is_pyobject
             self.is_temp = 1
+            if not self.type.is_pyobject:
+                original_type, self.type = self.type, py_object_type
+                # Hopefully this can be optimized out in some cases
+                return self.coerce_to(original_type, env)
+
         elif self.is_cpp_operation():
             self.analyse_cpp_operation(env)
         else:
             self.analyse_c_operation(env)
+        return self
 
     def is_py_operation(self):
         return self.is_py_operation_types(self.operand1.type, self.operand2.type)
@@ -12276,8 +12280,7 @@ class MulNode(NumBinopNode):
             elif operand2.type in builtin_sequence_types:
                 self.operand1 = operand1.coerce_to_index(env)
 
-        self.analyse_operation(env)
-        return self
+        return self.analyse_operation(env)
 
     @staticmethod
     def is_builtin_seqmul_type(type):
@@ -12414,7 +12417,7 @@ class DivNode(NumBinopNode):
 
     def analyse_operation(self, env):
         self._check_truedivision(env)
-        NumBinopNode.analyse_operation(self, env)
+        result = NumBinopNode.analyse_operation(self, env)
         if self.is_cpp_operation():
             self.cdivision = True
         if not self.type.is_pyobject:
@@ -12426,6 +12429,7 @@ class DivNode(NumBinopNode):
                 # Need to check ahead of time to warn or raise zero division error
                 self.operand1 = self.operand1.coerce_to_simple(env)
                 self.operand2 = self.operand2.coerce_to_simple(env)
+        return result
 
     def compute_c_result_type(self, type1, type2):
         if self.operator == '/' and self.ctruedivision and not type1.is_cpp_class and not type2.is_cpp_class:
@@ -12578,12 +12582,13 @@ class ModNode(DivNode):
             return "float divmod()"
 
     def analyse_operation(self, env):
-        DivNode.analyse_operation(self, env)
+        result = DivNode.analyse_operation(self, env)
         if not self.type.is_pyobject:
             if self.cdivision is None:
                 self.cdivision = env.directives['cdivision'] or not self.type.signed
             if not self.cdivision and not self.type.is_int and not self.type.is_float:
                 error(self.pos, "mod operator not supported for type '%s'" % self.type)
+        return result
 
     def generate_evaluation_code(self, code):
         if not self.type.is_pyobject and not self.cdivision:

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -11854,7 +11854,9 @@ class BinopNode(ExprNode):
             self.analyse_cpp_operation(env)
         else:
             self.analyse_c_operation(env)
-        return self
+        return self  # when modifying this function, remember that the
+            # DivNode and ModNode expect it to return either self, or something
+            # that wraps self - they relying on modifying self afterwards.
 
     def is_py_operation(self):
         return self.is_py_operation_types(self.operand1.type, self.operand2.type)
@@ -12424,8 +12426,6 @@ class DivNode(NumBinopNode):
         # The assumption here is that result is either 'self' or a coercion
         # node containing 'self'. Thus it is reasonable to keep manipulating
         # 'self' even if it's been replaced as the eventual result.
-        from . import Visitor
-        assert Visitor.tree_contains(result, self)
 
         if self.is_cpp_operation():
             self.cdivision = True
@@ -12438,7 +12438,7 @@ class DivNode(NumBinopNode):
                 # Need to check ahead of time to warn or raise zero division error
                 self.operand1 = self.operand1.coerce_to_simple(env)
                 self.operand2 = self.operand2.coerce_to_simple(env)
-        return result
+        return result  # should either be self, or wrap self
 
     def compute_c_result_type(self, type1, type2):
         if self.operator == '/' and self.ctruedivision and not type1.is_cpp_class and not type2.is_cpp_class:
@@ -12592,12 +12592,15 @@ class ModNode(DivNode):
 
     def analyse_operation(self, env):
         result = DivNode.analyse_operation(self, env)
+        # The assumption here is that result is either 'self' or a coercion
+        # node containing 'self'. Thus it is reasonable to keep manipulating
+        # 'self' even if it's been replaced as the eventual result.
         if not self.type.is_pyobject:
             if self.cdivision is None:
                 self.cdivision = env.directives['cdivision'] or not self.type.signed
             if not self.cdivision and not self.type.is_int and not self.type.is_float:
                 error(self.pos, "mod operator not supported for type '%s'" % self.type)
-        return result
+        return result  # should either be self, or wrap self
 
     def generate_evaluation_code(self, code):
         if not self.type.is_pyobject and not self.cdivision:

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -4989,13 +4989,14 @@ class MemoryViewIndexNode(BufferIndexNode):
         self.original_indices = indices
         self.nogil = env.nogil
 
-        self.analyse_operation(env, getting, axes)
+        self = self.analyse_operation(env, getting, axes)
         self.wrap_in_nonecheck_node(env)
         return self
 
     def analyse_operation(self, env, getting, axes):
         self.none_error_message = "Cannot index None memoryview slice"
         self.analyse_buffer_index(env, getting)
+        return self
 
     def analyse_broadcast_operation(self, rhs):
         """
@@ -5049,7 +5050,7 @@ class MemoryViewSliceNode(MemoryViewIndexNode):
 
         self.analyse_ellipsis_noop(env, getting)
         if self.is_ellipsis_noop:
-            return
+            return self
 
         self.index = None
         self.is_temp = True
@@ -5057,12 +5058,13 @@ class MemoryViewSliceNode(MemoryViewIndexNode):
 
         if not MemoryView.validate_axes(self.pos, axes):
             self.type = error_type
-            return
+            return self
 
         self.type = PyrexTypes.MemoryViewSliceType(self.base.type.dtype, axes)
 
         if not (self.base.is_simple() or self.base.result_in_temp()):
             self.base = self.base.coerce_to_temp(env)
+        return self
 
     def analyse_assignment(self, rhs):
         if not rhs.type.is_memoryviewslice and (

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -4989,9 +4989,9 @@ class MemoryViewIndexNode(BufferIndexNode):
         self.original_indices = indices
         self.nogil = env.nogil
 
-        self = self.analyse_operation(env, getting, axes)
-        self.wrap_in_nonecheck_node(env)
-        return self
+        node = self.analyse_operation(env, getting, axes)
+        node.wrap_in_nonecheck_node(env)
+        return node
 
     def analyse_operation(self, env, getting, axes):
         self.none_error_message = "Cannot index None memoryview slice"

--- a/tests/run/type_inference.pyx
+++ b/tests/run/type_inference.pyx
@@ -187,6 +187,8 @@ def arithmetic():
     assert typeof(f) == "long", typeof(f)
     g = 4 // <int>2
     assert typeof(g) == "long", typeof(g)
+    h = int(2) / 3.0
+    assert typeof(h) == "double", typeof(h)
 
 cdef class some_class:
     pass

--- a/tests/run/type_inference.pyx
+++ b/tests/run/type_inference.pyx
@@ -189,6 +189,12 @@ def arithmetic():
     assert typeof(g) == "long", typeof(g)
     h = int(2) / 3.0
     assert typeof(h) == "double", typeof(h)
+    i = 3.0 + int(5)
+    assert typeof(h) == "double", typeof(h)
+    j = int(-1) + 1.0
+    assert typeof(j) == "double", typeof(j)
+    k = 7.0 - int(2)
+    assert typeof(k) == "double", typeof(k)
 
 cdef class some_class:
     pass

